### PR TITLE
Align Checkbox Toggle implementation with description in Formfield co…

### PR DESCRIPTION
**Pull request by YueyingLu**
…ntext

### Description
To fix the misalignment between api description and implementatin:

**ariaDescribedby** and **ariaLabelledby**
> Adds aria-describedby to the component. If you're using this component within a form field, don't set this property because the form field component automatically sets it. 

> Adds aria-labelledby to the component. If you're using this component within a form field, don't set this property because the form field component automatically sets it. 

Toggle and Checkbox doesn't set description and label provided by parent Formfield(without this fix). They are necessary as the label and description of Formfield provides meaningful content to describe the usage of Toggle and Checkbox. For example, Formfield label and description act as group label or description when it wraps a group of checkbox.

**controlId**

> Specifies the ID of the native form element. You can use it to relate a label element's for attribute to this control. It defaults to an automatically generated ID that is provided by its parent form field component.

Toggle and Checkbox do not take controlId from parent Formfield in the implementation(with or without this fix). It makes sense because they usually have its own label, and no need to have Formfield label to control the form element in addition. Propose to change API description:

> Specifies the ID of the native form element. It uses an automatically generated ID by default.

Decision made in doc Checkbox-Toggle-and-Formfield-controlId-association

### How has this been tested?

[_How did you test to verify your changes?_]

[_How can reviewers test these changes efficiently?_]

[_Check for unexpected visual regressions, see [`CONTRIBUTING.md`](CONTRIBUTING.md#run-visual-regression-tests) for details._]

### Documentation changes

[*Do the changes include any API documentation changes?*]
- [ ] _Yes, this change contains documentation changes._
- [ ] _No._

### Related Links

[*Attach any related links/pull request for this change*]

<details>
   <summary>Review checklist</summary>

_The following items are to be evaluated by the author(s) and the reviewer(s)._

#### Correctness

- [ ] _Changes are backward-compatible if not indicated, see [`CONTRIBUTING.md`](CONTRIBUTING.md#public-apis)._
- [ ] _Changes do not include unsupported browser features, see [`CONTRIBUTING.md`](CONTRIBUTING.md#browsers-support)._
- [ ] _Changes were manually tested for accessibility, see [accessibility guidelines](https://cloudscape.design/foundation/core-principles/accessibility/)._

#### Security

- [ ] _If the code handles URLs: all URLs are validated through [the `checkSafeUrl` function](https://github.com/cloudscape-design/components/blob/main/src/internal/utils/check-safe-url.ts)._

#### Testing

- [ ] _Changes are covered with new/existing unit tests?_
- [ ] _Changes are covered with new/existing integration tests?_
</details>

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/checkbox/base-checkbox.tsx [hunk 1]
```diff
@@ -25,11 +25,7 @@ export interface BaseCheckboxProps extends BaseComponentProps, FormFieldControlP
   disabled?: boolean;
 
   /**
-   * Specifies the ID of the native form element. You can use it to relate
-   * a label element's `for` attribute to this control.
-   *
-   * It defaults to an automatically generated ID that
-   * is provided by its parent form field component.
+   * Specifies the ID of the native form element. It uses an automatically generated ID by default.
```
Comment (pan-kot) [inline src/checkbox/base-checkbox.tsx:28]: Why the "You can use it to relate a label element's `for` attribute to this control" part was removed?

File: src/checkbox/internal.tsx [hunk 2]
```diff
@@ -39,6 +38,7 @@ const InternalCheckbox = React.forwardRef<CheckboxProps.Ref, InternalProps>(
     },
     ref
   ) => {
+    const { ariaDescribedby, ariaLabelledby } = useFormFieldContext(rest);
```
Comment (pan-kot) [inline src/checkbox/internal.tsx:41]: Why using the context here and not inside AbstractSwitch?

If you do it inside the abstract switch it would also be possible to use the form-field controlId:

```
const formFieldContext = useFormFieldContext(rest);
const selfControlId = useUniqueId();
const controlId = propsControlId || formFieldContext.controlId || selfControlId;
```

**Review by pan-kot — CHANGES_REQUESTED** [review]

File: src/checkbox/base-checkbox.tsx (hunk 1 above)
Comment (YueyingLu) [reply src/checkbox/base-checkbox.tsx:28]: I think we don't need highlight it(it feels like recommendation), because unlike Input, Checkbox and Toggle usually have their own labels to control the form.

File: src/checkbox/internal.tsx (hunk 2 above)
Comment (YueyingLu) [reply src/checkbox/internal.tsx:41]: Inside AbstractSwitch may not work for RadioGroup -- we want to add ariaDescribedby and ariaLabelledby to RadioGroup but AbstractSwitch is for individual RadioButton.

File: src/checkbox/base-checkbox.tsx (hunk 1 above)
Comment (pan-kot) [reply src/checkbox/base-checkbox.tsx:28]: Sure, but as the controlId property exists I think it makes sense to describe its purpose fully. I'd imagine it is possible to provide an external label e.g. if you're building some selectable card or something alike.

If we want to discourage that use of the controlId for these elements I think it is better to deprecate this and provide an explanation.

File: src/checkbox/internal.tsx (hunk 2 above)
Comment (pan-kot) [reply src/checkbox/internal.tsx:41]: Can we then make the `controlId` property of the abstract switch required and compute this outside to still have the precedence `propsControlId || formFieldContext.controlId || selfControlId `? (for the radio group it is going to be selfControlId always).

File: src/checkbox/base-checkbox.tsx (hunk 1 above)
Comment (gethinwebster) [reply src/checkbox/base-checkbox.tsx:28]: can you include @stirache in this too, as it impacts public-facing documentation

File: src/checkbox/internal.tsx (hunk 2 above)
Comment (YueyingLu) [reply src/checkbox/internal.tsx:41]: It is the same for controlId(miss it in the last reply), AbstractSwitch controlId is for the individual RadioButton `controlId={item.controlId}`  not for the RadioGroup.

Comment (pan-kot) [reply src/checkbox/internal.tsx:41]: Yes, so for the checkbox we can do:

```
function Checkbox(props) {
  const context = useFormFieldContext(props);
  const selfControlId = useUniqueId();
  const controlId = props.controlId || context.controlId || selfControlId;
  return <AbstractSwitch
    controlId={controlId}
    ariaLabelledby={context.ariaLabelledby}
    ariaDescribedby={context.ariaDescribedby}
    ...
  />
}
```

and for radio group:

```
function RadioGroup(props) {
  const context = useFormFieldContext(props);
  const selfControlId = useUniqueId();
  const controlId = props.controlId || context.controlId || selfControlId;
  return <div role="radio-group"
    ariaLabelledby={context.ariaLabelledby}
    ariaDescribedby={context.ariaDescribedby}
    ...
  >
    {items.map((...) => <RadioButton {...} />)}
  </div>;
}

function RadioButton(props) {
  const controlId = useUniqueId();
  return <AbstractSwitch controlId={controlId} ... />;
}
```

File: src/checkbox/base-checkbox.tsx (hunk 1 above)
Comment (YueyingLu) [reply src/checkbox/base-checkbox.tsx:28]: The controlId can be used to access the input, not only to relate a label. So I don't discourage using it. But I am not keen to state this usage specifically, because it doesn't feel like a very common use case as they have their own label already. To build selectable cards, users should use card instead.

Comment (YueyingLu) [reply src/checkbox/base-checkbox.tsx:28]: @gethinwebster I will include Rachel when we have consensus what to include in the description.

Comment (pan-kot) [reply src/checkbox/base-checkbox.tsx:28]: I think the controlId is specifically meant for control/label association. Why would you need to assign an id on the native input otherwise? For testing we recommend using our test-utils that can be combined with date-testid.

Comment (YueyingLu) [reply src/checkbox/base-checkbox.tsx:28]: I dont have the data what else users may need the controlId. But for example we have controlId in Select, which is not meant (as I understand) for control/label association.

Comment (pan-kot) [reply src/checkbox/base-checkbox.tsx:28]: I think in select it works for the input/label association as well: https://github.com/cloudscape-design/components/blob/main/src/select/internal.tsx#L91.

```
// This line takes controlId from restProps or the form-field otherwise.
const formFieldContext = useFormFieldContext(restProps);
const selfControlId = useUniqueId('trigger');
const controlId = formFieldContext.controlId ?? selfControlId;
```

File: src/toggle/__tests__/toggle.test.tsx [hunk 3]
```diff
@@ -100,3 +103,55 @@ test('check a11y', async () => {
...
+  );
+  const toggleWrapper = createWrapper(container).findToggle()!;
+  const toggleAriaDescribedby = toggleWrapper.findNativeInput().getElement().getAttribute('aria-describedby');
+  const toggleAriaLabelledby = toggleWrapper.findNativeInput().getElement().getAttribute('aria-labelledby');
+
+  expect(toggleAriaDescribedby).toBe(
+    joinStrings('description-id', container?.querySelector(`.${abstractSwitchStyles.description}`)?.id)
+  );
+  expect(toggleAriaLabelledby).toBe(
+    joinStrings(container?.querySelector(`.${abstractSwitchStyles.label}`)?.id, 'label-id')
```
Comment (just-boris) [inline src/toggle/__tests__/toggle.test.tsx:155]: The implementation uses `joinStrings` already, so the test shouldn't, or it becomes `expect(add(2, 2)).toEqual(add(2, 2))` which is pointless.

Is there any reason not to put the full computed string here?

Comment (YueyingLu) [reply src/toggle/__tests__/toggle.test.tsx:155]: The point of the test is to check if toggle input has aria-describedby identified with formfield description id **AND** its own description id. So I think it is testing `expect(add(attr1, attr2)).toEqual(add(attr3, atrr4))`. joinString is not the point.

File: src/checkbox/base-checkbox.tsx (hunk 1 above)
Comment (YueyingLu) [reply src/checkbox/base-checkbox.tsx:28]: As I understand, the select takes the id from formfield, but not meant for input/label association? Also it is not stated in the description https://github.com/cloudscape-design/components/blob/main/src/select/interfaces.ts#L91

File: src/toggle/__tests__/toggle.test.tsx (hunk 3 above)
Comment (just-boris) [reply src/toggle/__tests__/toggle.test.tsx:155]: Having a static string in the right hand side is still preferred. If you cannot use a static string because there is a generated value, you can use `toMatch` ([example](https://github.com/cloudscape-design/components/blob/7433543fcd2796d6398b67867d25b637c1a2b6de/src/tabs/__tests__/tabs.test.tsx#L137))

```js
expect(toggleAriaLabelledby).toMatch(/^awsui-toggle-\d+ label-id/)
```

File: src/checkbox/base-checkbox.tsx (hunk 1 above)
Comment (pan-kot) [reply src/checkbox/base-checkbox.tsx:28]: In that case I would rather challenge the select.controlId description to include that info.

File: src/checkbox/internal.tsx (hunk 2 above)
Comment (pan-kot) [reply src/checkbox/internal.tsx:41]: Offline discussion: the control-id is not passed from form-field to the abstract-switch to avoid activating checkbox/radio/toggle when clicking on the form-field label. The current implementation addresses that.

File: src/checkbox/__tests__/checkbox.test.tsx [hunk 4]
```diff
@@ -155,3 +157,65 @@ test('does not trigger any change events when value is changed through api', ()
...
+  const checkboxInputAriaLabelledby = checkboxWrapper.findNativeInput().getElement().getAttribute('aria-labelledby');
+
+  const toggleLabelId = container?.querySelector(`.${abstractSwitchStyles.label}`)?.id;
+  const toggleDescriptionId = container?.querySelector(`.${abstractSwitchStyles.description}`)?.id;
+
+  expect(checkboxInputAriaDescribedby).toBe('description-id' + ' ' + toggleDescriptionId);
+  expect(checkboxInputAriaLabelledby).toBe(toggleLabelId + ' ' + 'label-id');
+});
+
+test('clicking on formfield label area wont call onchange', () => {
```
[the remark this answers is not in the document]
Comment (YueyingLu) [reply src/checkbox/__tests__/checkbox.test.tsx:209]: When adding the controlId from formfield to checkbox, this test failed. So it does do the job? And seems jsdom did fix for checkbox https://github.com/jsdom/jsdom/issues/2367. 
But as you said when testing onFocus in input, it does not work 🤔 .

[the remark this answers is not in the document]
Comment (just-boris) [reply src/checkbox/__tests__/checkbox.test.tsx:209]: The challenge here is that passing test may have two reasons 1) checkbox has explicitly deactivated this 2) it does not work for any components (which is unexpected but all tests are green)

This is why we need to have at least 2 tests in place, one that demonstrates active behavior and another one to demonstrate inactive behavior (maybe more tests to cover other corner-case type of controls, like Select or RadioGroup)

[the remark this answers is not in the document]
Comment (just-boris) [reply src/checkbox/__tests__/checkbox.test.tsx:209]: If you find a way to confirm it in JSDOM, this is fine for me too, but as far as I am aware, it cannot ensure this feature

[the remark this answers is not in the document]
Comment (YueyingLu) [reply src/checkbox/__tests__/checkbox.test.tsx]: Added integration test for Input Checkbox and Toggle.

File: src/checkbox/base-checkbox.tsx (hunk 1 above)
Comment (YueyingLu) [inline src/checkbox/base-checkbox.tsx:28]: @stirache please review this description change, thank you!

**Review by pan-kot — DISMISSED** [review]

File: src/checkbox/__tests__/checkbox.test.tsx [hunk 5]
```diff
@@ -3,8 +3,10 @@
 import React, { useState } from 'react';
 import { render } from '@testing-library/react';
 import createWrapper, { CheckboxWrapper } from '../../../lib/components/test-utils/dom';
+import FormField from '../../../lib/components/form-field';
 import Checkbox, { CheckboxProps } from '../../../lib/components/checkbox';
 import styles from '../../../lib/components/internal/components/checkbox-icon/styles.selectors.js';
+import abstractSwitchStyles from '../../../lib/components/internal/components//abstract-switch/styles.css.js';
```
Comment (just-boris) [inline src/checkbox/__tests__/checkbox.test.tsx:9]: Double slash in this import. Even if this works, better to keep it clean

```suggestion
import abstractSwitchStyles from '../../../lib/components/internal/components/abstract-switch/styles.css.js';
```

**Review by just-boris — COMMENTED** [review]
All good, except that comment about slashes

File: src/checkbox/base-checkbox.tsx (hunk 1 above)
Comment (stirache) [reply src/checkbox/base-checkbox.tsx:28]: "Specifies the ID of the native form element. It uses an automatically generated ID by default." --> "Specifies the ID of the native form element. By default, it uses an automatically generated ID."

**Review by stirache — CHANGES_REQUESTED** [review]

**Review by pan-kot — APPROVED** [review]

**Review by just-boris — APPROVED** [review]
